### ISSUE TITLE
nan of BN running_mean and running_var when finetuning resnet on my own dataset

### ISSUE THREAD

**Issue by nicklhy** [pytorch/pytorch#1206]
Hi everyone,
I am finetuing resnet152 on my own dataset recently. The training phase is perfect that the accuracy increased from random to around 90%.

But unfortunately, there is a serious problem when I tried to evaluate it. The model keeps producing the same results for all samples because all BN runing_mean and runing_var parameters in resnet152 are nan. (I do really not know why these parameters changed to nan when training the model.)

In addition, if I remove `model.eval()`(run this model in training mode) and set a relatively large batch-size, the models can once again output reasonable results. But in this case, BN layers will not use the learned parameters on train data. They just compute mean and var in the current mini-batch.

Do anyone know some potential reasons for this nan problem?

**Comment by nicklhy** [comment]
BTW, I construct the network like
```
class CustomNet(torch.nn.Module):
    def __init__(self, backbone):
        self.backbone = backbone
        self.fc = torch.nn.Linear(2048, 100) # 100 classes
        torch.nn.init.normal(self.fc.weight, 0, math.sqrt(2.0/(2048*100)))
        torch.nn.init.constant(self.fc.bias, 0)

    def forward(self, data):
        x = self.backbone(data)
        x = x.view((-1, 2048))
        clf = self.fc(x)
        return clf

resnet152_params = torch.load('/path/to/resnet152.pth')
resnet152 = torchvision.models.resnet152()
resnet152.load_state_dict(resnet152_params)

backbone = torch.nn.Sequential(*list(resnet152.children())[:-1])

net = CustomNet(backbone)
```

**Comment by soumith** [comment]
if one of your input batches has all zeros or a single nan value then these values can turn into nan.

**Comment by nicklhy** [comment]
@soumith , I don't think all zeros/nan samples can produce this problem since I use clip gradient when training the model. The learned parameters will not updated too much each time.
Moreover, the training phase looks good and the model works well if I replace net.eval() with net.train().

**Comment by soumith** [comment]
if sample has nan then it doesn't matter, batchnorm acts on input statistics. learning rate or gradient norm is irrelevant

**Comment by soumith** [comment]
and once a running_mean is nan, it's always nan until you reset it

**Comment by nicklhy** [comment]
OK, I will add some lines to check nan of input data and bn parameters in the training loop. New results will be reported later.

**Comment by nicklhy** [comment]
@soumith , thank you so much and I just found out the exact bug. You are right, the nan comes from forwarding an input data with all zero/nan values.
I have a function which can infer the feature shape of backbone network when building my custom network. Something like:

```
backbone = ...
sym_data = torch.autograd.Variable(torch.Tensor(1, 3, height, width))
sym_feat = self.backbone(sym_data)
num_features = sym_feat.data.size()[1]
```
Thus, the bn parameters are already nan before the train loop really began.

Now, the nan values disappeared after changing `sym_data = torch.autograd.Variable(torch.Tensor(1, 3, height, width))` to `sym_data = torch.autograd.Variable(torch.rand(1, 3, height, width))`.

**Comment by vadimkantorov** [comment]
If it's easy to protect running stats updates against NaN, it may be useful. Since a naive check:
`if torch.isnan(loss): continue` fails because of this :(
